content/blog/2019-12-18-heterocl-stream/index.md

```diff
@@ -0,0 +1,158 @@
+
+### Evaluation
+
+In this section, we evaluate our implementation by using both unit tests and realistic benchmarks. Experiments are performed on a server with 2.20GHz Intel Xeon processor and 128 GB memory. We verify the correctness of the final result and compare the total run time in different cases.
+
+#### Unit Tests
+
+The tests can be found [here](https://github.com/Hecmay/heterocl/blob/stream/tests/test_cpu_stream.py). Following we breifly illustrate what each test does by using the DFGs. 
+
+<img src="unit_test.png" width="700" >
+
+For unit tests, we compare the run time before and after applying data streaming. The results are shown in the following table. We run the results for 1000 times and calculate the average.
+
+| Testcase | Original (ms) | Multi-threading (ms) | Speedup |
+|:---------:|------------|------------|-----------|
+|two_stages|0.0592|0.0554| 1.070 |
+|three_stages|0.0831|0.0715| 1.162 |
+|internal_stage|N/A|0.0638| N/A |
+|fork_stage|0.0865|0.0758| 1.141 |
+|merge_stage|0.0906|0.0739| 1.226 |
+
+The average speedup of our test cases is 1.150, which makes sense because now we use multi-thread execution. Note that for the third benchmark (i.e., ``test_internal_stage``), the functionalities are different before and after applying data streaming. To be more specific, we list the test program here.
+
+```python
+@hcl.def_([A.shape, B.shape, C.shape, D.shape])
+def M1(A, B, C, D):
+    with hcl.for_(0, 10) as i:
+        B[i] = A[i] + 1
+        D[i] = C[i] + 1
+
+@hcl.def_([B.shape, C.shape])
+def M2(B, C):
+    with hcl.for_(0, 10) as i:
+        C[i] = B[i] + 1
+
+M1(A, B, C, D)
+M2(B, C)
+
+s = hcl.create_schedule([A, B, C, D])
+s.to(B, s[M2], s[M1], depth=1)
+s.to(C, s[M1], s[M2], depth=1)
+```
+
+We can see that without applying streaming, the production of ``D`` is not affected by ``M2``. However, if we specify ``C`` to be streamed from ``M2`` to ``M1``, the original memory read of ``C`` in ``M1`` now becomes a blocking read. This also demonstrates that without the simulation support for streaming, some hardware behaviors cannot be correctly represented.
+
+#### Realistic Benchmark
+
+We also show the evaluation results from a realistic benchmark, which is more complicated than the synthetic tests in the unit tests. Due to time limitation, we only use the Sobel edge detector, which is a popular edge detecting algorithm in image processing. We compare the results with the software simulation tool provided by the HLS compiler. More specifically, we first generate Vivado HLS code with ``hls::stream``. Then we use ``csim`` to run the software simulation. The evaluation results are shown below. We also show the time overhead due to compilation.
+
+| Simulation Method | Simulation Time (s) | Compilation Overhead (s) | Total Run Time (s) |
+|:---------:|------------|------------|--------|
+|LLVM JIT|0.00094|0|0.00094|
+|Vivado HLS csim|1.63|1.29|1.92|

```

Comment:
Hmm, it seems like the simulation time + compilation overhead should equal the total run time. But 1.63 + 1.29 ≠ 1.92.